src/benchmarks/fragmentation.cpp

```diff
@@ -0,0 +1,985 @@
+
+	if (args->is_poolset || util_file_is_device_dax(args->fname)) {
+		if (args->fsize < fb->poolsize) {
+			fprintf(stderr, "file size too large\n");
+			goto free_fb;
+		}
+		fb->poolsize = 0;
+	}
+	else if (fb->poolsize < PMEMOBJ_MIN_POOL) {
+		fb->poolsize = PMEMOBJ_MIN_POOL;
+	}
+
+	fb->poolsize = PAGE_ALIGNED_UP_SIZE(fb->poolsize);
+
+	fb->pop = pmemobj_create(args->fname, nullptr, fb->poolsize, args->fmode);
+	if (fb->pop == nullptr) {
+		fprintf(stderr, "%s\n", pmemobj_errormsg());
+		goto free_fb;
+	}
+
+	fb->pmemobj_array = (struct frag_obj **)malloc(sizeof(struct frag_obj*) *
+		n_objs);
+	if (fb->pmemobj_array == nullptr)
+	{
+		perror("malloc");
+		goto free_pop;
+	}
+	for (unsigned i = 0; i < n_objs; ++i)
+	{
+		fb->pmemobj_array[i]= (struct frag_obj *)
+			malloc(sizeof(struct frag_obj));
+
+		if (fb->pmemobj_array[i] == nullptr)
+		{
+			perror("malloc");
+			if(i > 0) {
+				fb->free_pmemobj_array(i - 1);
+			}
+			goto free_pop;
+		}
+		fb->pmemobj_array[i]->init(fb->pa);
+	}
+
+	fb->background_mem_usage =
+		parse_memory_usage_type(fb->pa->background_mem_usage_type_str);
+	scenario_index = parse_scenario(fb->pa->scenario);
+	if (scenario_index == -1) {
+		perror("wrong scenario name");
+		goto free_pop;
+	}
+	fb->func_op = scenarios[scenario_index].func_op;
+	fb->theoretical_memory_usage = 0;
+
+	pmembench_set_priv(bench, fb);
+	fragmentation = 0;
+	return 0;
+
+free_pop:
+	pmemobj_close(fb->pop);
+free_fb:
+	free(fb);
+	return -1;
+}
+
+/*
+ * frag_exit -- function for de-initialization benchmark
+ */
+static int
+frag_exit(struct benchmark *bench, struct benchmark_args *args)
+{
+	auto *fb = (struct frag_bench *)pmembench_get_priv(bench);
+
+	size_t n_ops = args->n_ops_per_thread *args->n_threads;
+	PMEMoid oid;
+	size_t remaining = 0;
+	size_t chunk = 100; /* calc at chunk level */
+	while (pmemobj_alloc(fb->pop, &oid, chunk, 0, NULL, NULL) == 0)
+		remaining += pmemobj_alloc_usable_size(oid) + 16;
+
+	size_t allocated_sum = 0;
+	size_t allocated_sum2 = 0;
+	oid = pmemobj_root(fb->pop, 1);
+	for (size_t n = 0; n < n_ops; ++n) {
+		if (fb->pmemobj_array[n]->is_allocated == false)
+			continue;
+		oid = fb->pmemobj_array[n]->oid;
+		oid.pool_uuid_lo = fb->pmemobj_array[n]->oid.pool_uuid_lo;
+		allocated_sum += pmemobj_alloc_usable_size(oid);
+		allocated_sum2 += pmemobj_alloc_usable_size(oid) +16;
+	}
+
+	size_t used = fb->poolsize - remaining;
+
+	fragmentation = ((float)used / fb->theoretical_memory_usage) - 1.f;
+	fragmentation2 = ((float)used / allocated_sum) - 1.f;
+	fragmentation3 = ((float)used / allocated_sum2) - 1.f;
+
+	fb->free_pmemobj_array((unsigned) n_ops);
+	pmemobj_close(fb->pop);
+	free(fb);
+	return 0;
+}
+
+/*
+* frag_print_fragmentation -- function to print additional information
+*/
+static void
+frag_print_fragmentation(struct benchmark *bench,
+	struct benchmark_args *args,
+	struct total_results *res)
+{
+	printf(";fragmentation:%f;%f;%f", fragmentation, fragmentation2,
```

Comment:
This will get printed only at the end of the benchmark. We need something that prints those values *during* the run, so that we can create a chart of fragmentation over time.
The intention is to run this for weeks at a time. Getting a single fragmentation number at the end is of little use.